In [1]:
run p1.py

**********
dt1: 1.0 tf1: 36000.0 df1: 324310917.977126
dt2: 0.5 tf2: 36000.0 df2: 324813020.5007068
delta t: 0.5
final distance: 324813020.5007068
error: -502102.52358078957
relative error: -0.0015458201854309502
**********
dt1: 0.25 tf1: 36000.0 df1: 325063935.01676553
dt2: 0.125 tf2: 36000.0 df2: 325189358.1714124
delta t: 0.125
final distance: 325189358.1714124
error: -125423.15464687347
relative error: -0.00038569267872769984
**********
dt1: 0.0625 tf1: 36000.0 df1: 325252061.2334843
dt2: 0.03125 tf2: 36000.0 df2: 325283410.63699913
delta t: 0.03125
final distance: 325283410.63699913
error: -31349.403514802456
relative error: -9.637566039230604e-05


The smallest delta t I could calculate is as follows (it took a while)<br>
dt1: 0.015625 tf1: 36000.0 df1: 325299084.8071171 <br>
dt2: 0.0078125 tf2: 36000.0 df2: 325306921.75922555 <br>
delta t: 0.0078125<br>
final distance: 325306921.75922555<br>
error: -7836.952108442783<br>
relative error: -2.409094791485337e-05<br>

That's<br>
$\Delta t = 0.0078125$<br>
$d(36000) = 325306921.75922555$<br>

I cut $\Delta t$ in half every time and noted the difference from the previous answer as error. Notice that $error$ and $\Delta t$ are being cut in half every time (since error is linear, $\Delta t$ and $error$ are directly proportional, so this makes sense). Thus,<br>
$\frac{|error_0|}{2^n} < 1000 \implies log_2(\frac{error_0}{1000}) < n$<br>
$error_0 = -502102.5$<br>
$\implies log_2(\frac{error_0}{1000}) = 8.97 < 9 = n$<br>
In order to achieve an accuracy less than 1000 meters, we need<br>
$\Delta t = \frac{\Delta t_0}{2^n}$ where $\Delta t_0 = 0.5$<br>
$\implies \Delta t = \frac{.5}{2^9} = \frac{1}{2^{10}} = \frac{1}{1024} < 0.001$<br>


In [1]:
run p1_b.py

Expected value: 325306921.75922555
dt: 200.0 d1(36000): 72414136.96110632 err: -252892784.79811925
dt: 210.0 d2(36000): 250667465.28667754 err: -74639456.47254801
d1 more accurate than d2: False
dt: 100.0 d1(36000): 217164438.65054289 err: -108142483.10868266
dt: 105.0 d2(36000): 385529825.5312745 err: 60222903.77204895
d1 more accurate than d2: False
dt: 66.66666666666667 d1(36000): 255035831.8619846 err: -70271089.89724094
dt: 70.0 d2(36000): 434429513.1276867 err: 109122591.36846113
d1 more accurate than d2: True
dt: 50.0 d1(36000): 273247626.2796922 err: -52059295.479533374
dt: 52.5 d2(36000): 242282579.4292694 err: -83024342.32995614
d1 more accurate than d2: True
dt: 40.0 d1(36000): 283968596.50312746 err: -41338325.25609809
dt: 42.0 d2(36000): 293943303.1236355 err: -31363618.635590076
d1 more accurate than d2: False
dt: 33.333333333333336 d1(36000): 291031076.07875544 err: -34275845.68047011
dt: 35.0 d2(36000): 329030518.6313102 err: 3723596.872084677
d1 more accurate than d2: 

The above output is messy, but it compares the computed d(36000) with the very small time step from (a). If error was smaller, it converges faster. d1 had $\Delta t$ divide $600$ but d2 did not. However, it was fairly even  in terms of which converged quicker (the former 15 times and the latter 14). 

Despite that, it makes sense why we would want our time steps to step on discontinuities. Around the discontinuity, it uses the left side of the discontinuity to predict past it. Any portion of delta t going over the discontinuity is being modeled incorrectly by the left side. If it steps on the discontinuity, it minimizes the extrapolation based on incorrect values.

# Problem 2
